libs/blockscout-client/src/client.rs

```diff
@@ -0,0 +1,112 @@
+use super::{config, Error, Result};
+use reqwest_middleware::ClientWithMiddleware;
+use url::Url;
+
+#[derive(Clone, Debug)]
+pub struct Client {
+    chain_id: String,
+    base_url: Url,
+    api_sensitive_endpoints_key: Option<String>,
+    request_client: ClientWithMiddleware,
+}
+
+impl Client {
+    pub fn new(config: config::Config) -> Self {
+        let config = match config::ValidatedConfig::try_from(config) {
+            Ok(config) => config,
+            Err(err) => panic!("Invalid blockscout client configuration: {err}"),
+        };
+
+        let request_client = {
+            let client = reqwest::Client::new();
+            let mut client_with_middleware = reqwest_middleware::ClientBuilder::new(client);
+            for middleware in config.middleware_stack {
+                client_with_middleware = client_with_middleware.with_arc(middleware);
+            }
+            client_with_middleware.build()
+        };
+
+        Self {
+            chain_id: config.chain_id,
+            base_url: config.url,
+            api_sensitive_endpoints_key: config.api_sensitive_endpoints_key,
+            request_client,
+        }
+    }
+
+    pub fn chain_id(&self) -> &str {
+        self.chain_id.as_ref()
+    }
+
+    pub fn api_sensitive_endpoints_key(&self) -> Option<&str> {
+        self.api_sensitive_endpoints_key.as_deref()
+    }
+
+    pub fn build_url(&self, path: &str) -> Url {
+        self.build_url_with_query(path, None)
+    }
+
+    pub(crate) fn build_url_with_query(&self, path: &str, query: Option<&str>) -> Url {
+        let mut url = self.base_url.clone();
+        url.set_path(path);
+        url.set_query(query);
+        url
+    }
+
+    pub(crate) async fn _post_request<U, Request, Response>(
```

Comment:
why _ ?

Reply:
We do not use it for now and as it is not public it should be used somewhere in order not to cause the compilation warning. However it was just copied from the client implementation of `smart-contract-verifier`, and as we will extend that client further, I believe it makes sense to leave it so it can be easily used when required